docs/coding-guidelines/mono-code-guide.md

```diff
@@ -0,0 +1,253 @@
+
+* A non-public-API function should take a `MonoError *` argument.
+* In case of an error the function should call one of the `mono_error_set_` functions from [../../src/mono/mono/utils/mono-error-internals.h](../../src/mono/mono/utils/mono-error-internals.h)
+* Inside the runtime check if there was an error by calling `is_ok (error)`, `mono_error_assert_ok (error)`, `goto_if_nok` or `return_if_nok`/`return_val_if_nok`
+* If there is an error and you're handling it, call `mono_error_cleanup (error)` to dispose of the resources.
+* `MonoError*` is generally one-shot: after it's been cleaned up it need to be re-inited with `mono_error_init_reuse`, but this is discouraged.
+* Instead if you intend to deal with an error, use `ERROR_DECL (local_error)` then call the
+  possibly-failing function and pass it `local_error`, then call `mono_error_assert_ok
+  (local_error)` if it "can't fail" or `mono_error_cleanup (local_error)` if you want to ignore the
+  failure.
+
+## Managed Exceptions
+
+New code should generally not deal with `MonoException*`, use `MonoError*` instead.
+
+New code should avoid using `mono_error_set_pending_exception` - it affects a thread-local flag in
+a way that is not obvious to the caller of your function and may trample existing in-flight
+exceptions and make your code fail in unexpected ways.
+
+There are two circumstances when you might need to call `mono_error_set_pending_exception`:
+
+1. You're working with a public Mono API that sets a pending exception
+2. You're implementing an icall, but can't use `HANDLES()` (see the [internal calls](#internal-calls) section below)
+
+## Internal calls <a name="internal-calls"></a>
+
+Prefer P/Invokes or QCalls over internal calls.  That is, if your function only takes arguments that
+are not managed objects, and does not need to interact with the runtime, it is better to define it
+outside the runtime.
+
+Internal calls generally have at least one argument that is a managed object, or may throw a managed exception.
+
+Internal calls are declared in [`icall-def.h`](../../src/mono/mono/metadata/icall-def.h) See the comment in the header for details.
+
+There are two styles of internal calls: `NOHANDLES` and `HANDLES`.  (This is a simplification there
+are also JIT internal calls added by the execution engines)
+
+The difference is that `HANDLES` icalls receive references to managed objects wrapped in a handle
+that attempts to keep the object alive for the duration of the internal call even if the thread
+blocks or is suspended, while `NOHANDLES` functions don't.  Additionally `HANDLES` functions get a
+`MonoError*` argument from the managed-to-native interop layer that will be converted to a managed
+exception when the function returns.  `NOHANDLES` functions generally have to call
+`mono_error_set_pending_exception` themselves.
+
+## Suspend Safety
+
+See [Cooperative Suspend at mono-project.com](https://www.mono-project.com/docs/advanced/runtime/docs/coop-suspend/).
```

Comment:
Should these docs be in this repo, like the BOTR for CoreCLR is?

Reply:
Yeah I think we should eventually migrate them here